### PR TITLE
Image v4

### PR DESCRIPTION
**Pull request by pngwn**
## Description

Please include a concise summary, in clear English, of the changes in this pull request. If it closes an issue, please mention it here.

Closes: #(issue)

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [inline client/python/gradio_client/client.py:874]: Replace `FILE_DATA` in the client utils with `FILE_DATA = "'Dict(path: str, url: str | None, size: int | None, orig_name: str | None, mime_type: str | None)'"`

Will think of a better way to do this going forward

File: gradio/imagetools.py
Comment (freddyaboulton) [inline gradio/imagetools.py:1]: add `from __future__ import annotations` or else we can't import in 3.8 because of `|`

File: gradio/data_classes.py
Comment (freddyaboulton) [inline gradio/data_classes.py:122]: Also change `serialize_data` in blocks to be 

```python
        def format_file(s):
            return FileData(path=s).model_dump()
```

File: client/js/package.json
Comment (pngwn) [inline client/js/package.json:16]: change to looser pin

File: client/js/src/client.ts
Comment (pngwn) [inline client/js/src/client.ts:866]: remove

File: client/python/gradio_client/client.py
Comment (pngwn) [inline client/python/gradio_client/client.py:944]: Do we need to check for the version here and do both? I don't think this is backwards compatible.

File: client/python/gradio_client/client.py
Comment (pngwn) [inline client/python/gradio_client/client.py:1026]: I swapped this around because we don't have `is_file` any more, `is_file` is implicit. For previous version i think the `data` check should work, if it is present it should be base64 data. New versions that do not have a data key will skip this branch.

File: client/python/gradio_client/data_classes.py
Comment (pngwn) [inline client/python/gradio_client/data_classes.py:10]: I left this on the client `FileData` model for the above backwards compat reasons.

File: client/python/gradio_client/utils.py
Comment (pngwn) [inline client/python/gradio_client/utils.py:663]: remove

File: gradio/__init__.py
Comment (pngwn) [inline gradio/__init__.py:95]: So we don't actually have the ImageEditor in this PR, so these templates don't make much sense. I will bring them back but they might not make a huge amount of sense in v4. We can see.

I don't think the `Webcam` templates really makes any sense now so I did remove that one. I also think `Pil` is a little excessive.

File: gradio/blocks.py
Comment (pngwn) [inline gradio/blocks.py:1111]: remove

File: gradio/cli/commands/components/build.py
Comment (pngwn) [inline gradio/cli/commands/components/build.py:58]: remove, wrong branch

File: gradio/components/image.py
Comment (pngwn) [inline gradio/components/image.py:57]: ruff doesn't like lists as a default argument

File: gradio/components/image.py
Comment (pngwn) [inline gradio/components/image.py:127]: new streaming check.

File: gradio/components/image.py
Comment (pngwn) [inline gradio/components/image.py:169]: I move this into a utils file because i'll need it for `ImageEditor` too and it cleans up the component a bit. Had to cast cos the typing sucks

File: gradio/components/image.py
Comment (pngwn) [inline gradio/components/image.py:183]: Same as above with saving images, made it a util for reuse.

File: gradio/components/upload_button.py
Comment (pngwn) [inline gradio/components/upload_button.py:126]: type was wrong here

File: gradio/components/upload_button.py
Comment (pngwn) [inline gradio/components/upload_button.py:144]: same

File: gradio/components/video.py
Comment (pngwn) [inline gradio/components/video.py:156]: remove

File: gradio/processing_utils.py
Comment (pngwn) [inline gradio/processing_utils.py:276]: I think this is correct but not certain

File: gradio/templates.py
Comment (pngwn) [inline gradio/templates.py:1]: I've mostly just commented these out for now.

File: js/atoms/src/IconButton.svelte
Comment (pngwn) [inline js/atoms/src/IconButton.svelte:8]: Change to `bordered`

File: js/atoms/src/UploadText.svelte
Comment (pngwn) [inline js/atoms/src/UploadText.svelte:33]: remove

File: js/upload/src/utils.ts
Comment (pngwn) [inline js/upload/src/utils.ts:18]: Renamed these args for clarity.

File: js/upload/src/utils.ts
Comment (pngwn) [inline js/upload/src/utils.ts:162]: `FileData` is a class now, felt a little simpler and better parity with the backend.

File: test/test_components.py
Comment (pngwn) [inline test/test_components.py:607]: I didn't think we needed all of these base64 tests but could be wrong.

For some reason this test is not passing even thought it seems to be the same. WOuld love some help debugging.

File: test/test_routes.py
Comment (pngwn) [inline test/test_routes.py:314]: Changes this to upload a file instead, since the base64 thing gives validation errors now.

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [reply client/python/gradio_client/client.py:947]: It's ok - backwards compatibility is handled by another endpoint class entirely

File: client/python/gradio_client/client.py
Comment (pngwn) [reply client/python/gradio_client/client.py:1037]: @freddyaboulton does that mean I can get rid of this too?

File: gradio/components/image.py
Comment (freddyaboulton) [reply gradio/components/image.py:57]: Mutable default arguments are actually not good practice because they get shared across all instances of the class. 

I think we can:
* Set default to None and set sources to ["upload", "webcam", "clipboard"] if it's None
* Set the type to `Iterable[Literal` and use a tuple as the default argument.

Second might be nicer.

File: gradio/components/image.py
Comment (freddyaboulton) [reply gradio/components/image.py:127]: Shouldn't it be `not "webcam" in sources` since it could be in any location in the sources array?

File: gradio/components/image.py
Comment (freddyaboulton) [reply gradio/components/image.py:169]: Nice

File: gradio/processing_utils.py
Comment (freddyaboulton) [reply gradio/processing_utils.py:276]: Looks good to me tbh

File: js/upload/src/Upload.svelte
Comment (freddyaboulton) [inline js/upload/src/Upload.svelte:112]: The 🐍 -nification of javascript lol

File: js/upload/src/utils.ts
Comment (freddyaboulton) [inline js/upload/src/utils.ts:72]: I think this is wrong - path is never null since we set it in `upload` so we don't append the `/file=` part

**Review by freddyaboulton — COMMENTED** [review]
Backend changes look good to me @pngwn !! 

Since you said the front-end is still wip, I didn't take too close a look. The redesign is awesome though!! I love how the new image upload interface.

File: js/image/shared/ImageUploader.svelte
Comment (freddyaboulton) [inline js/image/shared/ImageUploader.svelte:186]: This should be url. Actually I think my comment about `normallise_file` above is wrong then.

File: gradio/components/image.py
Comment (pngwn) [reply gradio/components/image.py:57]: cool

File: gradio/components/image.py
Comment (pngwn) [reply gradio/components/image.py:127]: No, so the idea (for now at least) is that iit must be exactly 'webcam' for streaming to take effect because i doubt people want to handle streams of image snapshot as well as discreet images. Streaming also won't work with other sources types i think (maybe).

So if the sources == `['webcam', 'upload']` then this error should be raised.

File: js/image/shared/ImageUploader.svelte
Comment (pngwn) [reply js/image/shared/ImageUploader.svelte:186]: Yah, I've fixed this now and also got rid of the `_value` thing, we don't need it now.

File: js/upload/src/utils.ts
Comment (pngwn) [reply js/upload/src/utils.ts:72]: I still think it might be wrong, i thought that when i refactored but I didn't wanna change anything. I don't really understand when this would ever be useful, even if it is supported. Unless we use it to set some kind of base path for `sample_dir` in examples or something?

Comment (freddyaboulton) [reply js/upload/src/utils.ts:72]: I think this might be ok after all. I think images were not being loaded because we were using `url` instead of `path` as the img soruce

File: gradio/imagetools.py
Comment (abidlabs) [inline gradio/imagetools.py:1]: Consider calling this file `image_utils.py` for consistency

File: client/python/gradio_client/data_classes.py
Comment (abidlabs) [reply client/python/gradio_client/data_classes.py:10]: Let's comment to explain this is only here for backwards compatibility. 

Regarding this, why is `data` needed for backwards compatibility, but renaming `name` -> `path` doesn't break backwards compatibility?

File: client/python/gradio_client/client.py
Comment (abidlabs) [inline client/python/gradio_client/client.py:877]: Remove

File: gradio/components/image.py
Comment (abidlabs) [reply gradio/components/image.py:57]: We use the first approach in most other places, so maybe we should go with that. But don't feel too strongly.

File: gradio/components/image.py
Comment (abidlabs) [inline gradio/components/image.py:55]: Once we create the `ImageEditor` component, it'd be nice to add the `tool` argument back to `Image`, and if a person tries to create an `Image` with a tool, we display a warning and just create the `ImageEditor` component for them. I think it'll help with the migration

**Comment by abidlabs** [comment]
Awesome cleanup of the backend @pngwn! Tested it a few times, love the simplified component. Regarding the UI:

image

* For the upload icon, can we use the same one as in the Audio component for consistency?

image

* I also like the soft horizontal rule on top of the icons in the `Audio` component. For consistency, should we add it for the Image component? Not sure if it'll look good when there's an image uploaded so we should try it and see how it looks

* I like the paste icon a lot, but the webcam icon looks more like video than webcam. Could we use an icon like this instead?

image

* The "X" icon was handy to clear an uploaded image -- it'd be great to have that again for interactive mode.

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [reply client/python/gradio_client/client.py]: Sorry I missed this - what do you mean?

File: gradio/components/image.py
Comment (freddyaboulton) [reply gradio/components/image.py:127]: Ah ok. I think it would be better to rewrite as `if streaming and sources != ['webcam']` then!

File: client/python/gradio_client/client.py
Comment (pngwn) [reply client/python/gradio_client/client.py]: This is for backwards compat too, was referring to the above comment as backwards compat being handled by another class.

File: client/python/gradio_client/data_classes.py
Comment (pngwn) [reply client/python/gradio_client/data_classes.py:10]: @freddyaboulton Can i remove this data field as per the above comment or do I need to add `path` back?

**Comment by pngwn** [comment]
Will address these comments and get this merged to unblock stuff, we can continue to tweak after that if needed.

File: gradio/components/image.py
Comment (pngwn) [reply gradio/components/image.py:55]: Sounds good.

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [inline client/python/gradio_client/client.py:1026]: This function is only called on `V4` spaces so it's ok to delete the data check since data will no longer be present in V4 after this is merged.

Don't worry if client tests fail as long as they're connecting to a space on spaces. It's cause they're running an older version of the client.

File: client/python/gradio_client/data_classes.py
Comment (freddyaboulton) [reply client/python/gradio_client/data_classes.py:10]: This is only used for type hints in the `serializing.py` class, which is only needed for backwards compatibility with v3.

So let's change `path` back to `name` and make it clear this is only used in V3 spaces.

**Comment by pngwn** [comment]
@abidlabs I've added back the clear icon and changed the Upload and Webcam icons. I added a soft border above the icons too. It looked weird with an image so it goes away when an image is uploaded but is there when the image is empty.

Will sort out conflicts and tests and get this merged. Feel free to rereview before or after it is merged and we can address any issues.

**Comment by abidlabs** [comment]
Small UI bug: if the length of the webcam is too long, it appears cut off like this:

image

Otherwise, UI looks super nice! Doing some more testing with complete apps

File: gradio/components/image.py
Comment (abidlabs) [inline gradio/components/image.py:115]:
```suggestion
        if isinstance(sources, str) and sources in valid_sources:
            sources = [sources]
        for source in sources:
```

File: gradio/components/image.py
Comment (abidlabs) [inline gradio/components/image.py:114]:
```suggestion
        if isinstance(sources, str):
            sources = [sources]
```

File: gradio/components/image.py
Comment (abidlabs) [inline gradio/components/image.py:118]:
```suggestion
                    f"`sources` must a list consisting of elements in {valid_sources}"
```

**Comment by abidlabs** [comment]
Tested a bunch of demos, and generally everything is looking good. We need to update `source` -> `sources` in some of the demos, such as `kitchen_sink` but that can be done in a separate PR. The only thing that I noticed that wasn't working was examples. 

E.g. 

```py
import gradio as gr

gr.Interface(lambda x:x, gr.Image(), gr.Image(), examples=["cheetah.jpg"]).launch()
```

Clicking on the example image produced a collapsed input image:

image

**Review by abidlabs — APPROVED** [review]
Awesome redesign @pngwn! Approving as the few things I noticed can be done in a separate PR

**Comment by pngwn** [comment]
Thanks for the pointers and reviews @abidlabs & @freddyaboulton! Much appreciated.

I've fixed the examples which is the biggest issue, i'll leave the dropdown for another PR and merge this in when CI is good.

**Comment by pngwn** [comment]
I have had to comment out a few tests because we don't have the ImageEditor atm. I'll add those back and fix the demos when ImageEditor has returned.

**Comment by abidlabs** [comment]
Sounds great @pngwn thanks for addressing the review!

File: js/audio/static/StaticAudio.svelte
Comment (freddyaboulton) [inline js/audio/static/StaticAudio.svelte:46]: @pngwn here we're using `url` but in `VideoPreview` we're using `data`. Which is intended?

**Pull request merged**
